documentation/frontend/development.md

```diff
@@ -0,0 +1,215 @@
+
+### 🚀 Production version
+
+The `make dev` command runs the `docker-compose.yml` which runs the `dev` target in the [Dockerfile](./Dockerfile). To run a production version in docker, run `docker compose up -d -f docker-compose-realease.yml` which targest the `release` stage in the docker build. This runs the production version, while still creating a network connection to the local API.
+
+### Testing Release Target Locally
+
+To test the release target locally, run:
+
+- `make release-build OPTS="--tag [IMAGE_NAME]"` or
+- `docker buildx build --target release --tag [IMAGE_NAME]` for a faster build on OSX
+
+to build a local image. To view the site at `localhost:3000`, run: `docker run -e "HOSTNAME=0.0.0.0" -p 3000:3000 [IMAGE_NAME]`.
+
+## 🎯 Testing
+
+### :atom: Unit Testing
+
+[Jest](https://jestjs.io/docs/getting-started) is used as the test runner and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) provides React testing utilities.
+
+Tests are manged as `.test.ts` (or `.tsx`) files in the the `tests/` directory.
+
+To run tests:
+
+- `npm test` - Runs all tests and outputs test coverage report
+- `npm run test-update` - Updates test snapshots
+- `npm run test-watch` - Runs tests in [watch](https://jestjs.io/docs/cli#--watch) mode. Tests will re-run when files are changed, and an interactive prompt will allow you to run specific tests or update snapshots.
+
+A subset of tests can be ran by passing a pattern to the script. For example, to only run tests in `tests/pages/`:
+
+```sh
+npm run test-watch -- pages
+```
+
+### 🚦 End-to-end (E2E) testing
+
+[Playwright](https://playwright.dev/) is a framework for web testing and its test runner is called [Playwright Test](https://playwright.dev/docs/api/class-test), which can be used to run E2E or integration tests across chromium, firefox, and webkit browsers.
+
+E2E test filenames end with `.spec.ts` and are found in the `tests/e2e` directory.
+
+To run E2E tests via CLI:
+
+- `cd ../api && make init db-seed-local start` (prerequisite to start the API)
+- `npx playwright install --with-deps` — Downloads playwright browsers required to run tests
+- `npm run test:e2e` — Runs all E2E tests using the playwright config found at `tests/playwright.config.ts`
+- `npm run test:e2e:ui` — Run specific or all E2E tests using Playwright's [UI mode](https://playwright.dev/docs/test-ui-mode), which is useful for debugging full traces of each test
+
+To run E2E tests using VS Code:
+
+1. Download the VS Code extension described in these [Playwright docs](https://playwright.dev/docs/running-tests#run-tests-in-vs-code)
+2. Follow the [instructions](https://playwright.dev/docs/getting-started-vscode#running-tests) Playwright provides
+
+Playwright E2E tests run "local-to-local", requiring both the frontend and the API to be running for the tests to pass - and for the database to be seeded with data.
```

Comment:
I've been confused by this - I actually get an error if I try to run e2e tests locally in VS code while running a dev server. If I have a built application in the .next directory, though, it seems to work without starting an actual server, and seems to start its own automatically

Reply:
since it's running `npm run start` on 3000, it seems like what you actually need to do is make sure you have a build complete for playwright to use, which would mean running `npm run build` first. That tracks with my experience, but the docs here seem to suggest that you actually need to have things running (either with `npm start` or `npm run dev`) first, which in my experience doesn't work

Reply:
`reuseExistingServer: !process.env.CI,` means that it [just listens on 3000](https://playwright.dev/docs/test-webserver).

So you need to run on dev or prod build locally.